src/VerifyEmailHelper.php

```diff
@@ -63,21 +64,60 @@ public function generateSignature(string $routeName, string $userId, string $use
         return new VerifyEmailSignatureComponents(\DateTimeImmutable::createFromFormat('U', (string) $expiryTimestamp), $signature, $generatedAt);
     }
 
-    public function validateEmailConfirmation(string $signedUrl, string $userId, string $userEmail): void
+    public function validateEmailConfirmation(string $signedUrl, string $userId, string $userEmail, ?Request $request = null): void
     {
-        if (!$this->uriSigner->check($signedUrl)) {
+        if (null === $request) {
+            //@trigger_deprecation('You must pass a Request object....');
+        }
+
+        if ($hasRequestObject = $this->uriSigner instanceof UriSigner && null !== $request) {
+            $isValid = $this->uriSigner->checkRequest($request);
+        } else {
+            // @trigger_deprecation('Don\'t use a string anymore');
+            $isValid = $this->uriSigner->check($signedUrl);
+        }
+
+        if (!$isValid) {
             throw new InvalidSignatureException();
         }
 
-        if ($this->queryUtility->getExpiryTimestamp($signedUrl) <= time()) {
+        if ($hasRequestObject) {
+            $expiresAt = $request->query->getInt('expires');
+            $userToken = $request->query->getString('token');
+        } else {
+            $expiresAt = $this->queryUtility->getExpiryTimestamp($signedUrl);
+            $userToken = $this->queryUtility->getTokenFromQuery($signedUrl);
+        }
+
+        if ($expiresAt <= time()) {
             throw new ExpiredSignatureException();
         }
 
         $knownToken = $this->tokenGenerator->createToken($userId, $userEmail);
-        $userToken = $this->queryUtility->getTokenFromQuery($signedUrl);
 
         if (!hash_equals($knownToken, $userToken)) {
             throw new WrongEmailVerifyException();
         }
     }
+
+    public function validateEmailConfirmationFromRequest(Request $request, string $userId, string $userEmail): void
```

Comment:
Another option would be to pass a value object into this method instead of the 3 as shown. This would "future proof" the interface so we don't have to make changes to it again...

`validateEmailFromConfirmation(Confirmation $confirmation): void`

I suppose we could do this instead in `validateEmailConfirmation(string $signedUrl, string $userId, string $userEmail, ?Confirmation $confirmation = null)` _if_ we keep that method.